src/main/aerospike.c

```diff
@@ -83,187 +88,535 @@ static PyMethodDef Aerospike_Methods[] = {
+struct type_name_to_creation_method {
+    const char *name;
+    PyTypeObject *(*pytype_ready_method)(void);
+};
 
-    py_global_hosts = PyDict_New();
+static struct type_name_to_creation_method py_module_types[] = {
+    // We also don't retrieve the type's __name__ because:
+    // 1. Some of the objects have names different from the class name when accessed from the package
+    // 2. We don't want to deal with extracting an object's __name__ from a Unicode object.
+    // We have to make sure the Unicode object lives as long as we need its internal buffer
+    // It's easier to just use a C string directly
+    {"Client", AerospikeClient_Ready},
+    {"Query", AerospikeQuery_Ready},
+    {"Scan", AerospikeScan_Ready},
+    {"KeyOrderedDict", AerospikeKeyOrderedDict_Ready},
+    {"GeoJSON", AerospikeGeospatial_Ready},
+    {"null", AerospikeNullObject_Ready},
+    {"CDTWildcard", AerospikeWildcardObject_Ready},
+    {"CDTInfinite", AerospikeInfiniteObject_Ready},
+};
 
-    PyObject *exception = AerospikeException_New();
-    if (exception == NULL) {
+PyMODINIT_FUNC PyInit_aerospike(void)
+{
+    static struct PyModuleDef moduledef = {
+        PyModuleDef_HEAD_INIT,
+        .m_name = AEROSPIKE_MODULE_NAME,
+        .m_doc = "Aerospike Python Client",
+        .m_methods = aerospike_methods,
+    };
+
+    PyObject *py_aerospike_module = PyModule_Create(&moduledef);
+    if (py_aerospike_module == NULL) {
         return NULL;
     }
 
-    Py_INCREF(exception);
-    int retval = PyModule_AddObject(aerospike, "exception", exception);
-    if (retval == -1) {
-        goto CLEANUP;
-    }
-    Aerospike_State(aerospike)->exception = exception;
-
-    PyTypeObject *client = AerospikeClient_Ready();
-    Py_INCREF(client);
-    retval = PyModule_AddObject(aerospike, "Client", (PyObject *)client);
-    if (retval == -1) {
-        goto CLEANUP;
-    }
-    Aerospike_State(aerospike)->client = client;
+    Aerospike_Enable_Default_Logging();
 
-    PyTypeObject *query = AerospikeQuery_Ready();
-    Py_INCREF(query);
-    retval = PyModule_AddObject(aerospike, "Query", (PyObject *)query);
-    if (retval == -1) {
-        goto CLEANUP;
+    py_global_hosts = PyDict_New();
+    if (py_global_hosts == NULL) {
+        goto MODULE_CLEANUP_ON_ERROR;
     }
-    Aerospike_State(aerospike)->query = query;
 
-    PyTypeObject *scan = AerospikeScan_Ready();
-    Py_INCREF(scan);
-    retval = PyModule_AddObject(aerospike, "Scan", (PyObject *)scan);
-    if (retval == -1) {
-        goto CLEANUP;
+    unsigned long i = 0;
+    int retval;
+    for (i = 0; i < sizeof(py_submodules) / sizeof(py_submodules[0]); i++) {
+        PyObject *(*create_py_submodule)(void) =
+            py_submodules[i].pyobject_creation_method;
+        PyObject *py_submodule = create_py_submodule();
+        if (py_submodule == NULL) {
+            goto GLOBAL_HOSTS_CLEANUP_ON_ERROR;
+        }
+
+        retval = PyModule_AddObject(py_aerospike_module, py_submodules[i].name,
```

Comment:
Why no Py_INCREF(py_submodule) when py_type's reference count is increased, what's the difference?

Reply:
When the `exception` and `predicates` submodules are created, a strong reference is returned.

https://github.com/aerospike/aerospike-client-python/blob/15.1.0/src/main/exception.c#L269
https://github.com/aerospike/aerospike-client-python/blob/15.1.0/src/main/predicates.c#L393
https://docs.python.org/3/c-api/module.html#c.PyModule_Create

PyModule_AddObject steals a reference to value only on success: https://docs.python.org/3/c-api/module.html#c.PyModule_AddObject

I incremented the reference count for each type object being added to `aerospike` because I was following the example code in this tutorial:
https://docs.python.org/3/extending/newtypes_tutorial.html